### ISSUE TITLE
Lint/RedundantSafeNavigation should ignore self-yielding methods

### ISSUE THREAD

**Issue by tsugimoto** [rubocop-hq/rubocop#8868]
I've just updated to RuboCop 0.93.0 and received some suspicious false positives from Lint/RedundantSafeNavigation.

Safe navigation usages are intended to "check `nil`, and we expect an early return with `nil` result when the receiver was `nil`, so we can ease `nil` handling."
Self-yielding methods, `#tap` , `#yield_self`, and `#then`, in the other hand, yields the receiver into a block when called, possibly making method calls with the possibly-`nil` receiver, or pass `nil` to unsafe methods like `Time.at`.
So safe navigating into self-yielding methods should imply that an early return is intended, therefore I think the safe navigations are actually not redundant.

While I can assume that auto-correction may break code with unsafe cops, currently the possible actions I can take are either rewrite safe navigation to something like `array_or_nil.tap { |array| puts array[0] unless array.nil? }`, which is even more redundant, or disable this cop.

--------

## Expected behavior

`#tap` , `#yield_self`, and `#then` should not be flagged.

## Actual behavior

These calls are flagged.
Auto-correction produces code that passes `nil` to a block which may raise errors if `nil` is passed.

While documentation says `it can produce code that returns non nil`, raising is more than expected.

## Steps to reproduce the problem

```
begin
  nil&.tap { |i| puts "p #{Time.at(i)}" }
rescue
  puts 'parse error'
end

time = begin
  nil&.yield_self { |i| Time.at(i) }
rescue
  'parse error'
end

puts "time = #{time.inspect}"
```

```
$ ruby main.rb
time = nil
$ rubocop --auto-correct-all --only Lint/RedundantSafeNavigation main.rb
Inspecting 1 file
W

Offenses:

main.rb:2:6: W: [Corrected] Lint/RedundantSafeNavigation: Redundant safe navigation detected.
  nil&.tap { |i| puts "p #{Time.at(i)}" }
     ^^^^^
main.rb:8:6: W: [Corrected] Lint/RedundantSafeNavigation: Redundant safe navigation detected.
  nil&.yield_self { |i| Time.at(i) }
     ^^^^^^^^^^^^

1 file inspected, 2 offenses detected, 2 offenses corrected
$ ruby main.rb
parse error
time = "parse error"
$
```

## RuboCop version

```
$ [bundle exec] rubocop -V
0.93.0 (using Parser 2.7.2.0, rubocop-ast 0.7.1, running on ruby 2.6.3 x86_64-darwin18)
```

**Comment by bbatsov** [comment]
> So safe navigating into self-yielding methods should imply that an early return is intended, therefore I think the safe navigations are actually not redundant.

Fair point. I guess we didn't think this through. //cc @fatkodima
